src/ddspdrum/module.py

```diff
@@ -305,20 +326,36 @@ def __call__(self, envelopecontrol: np.array, audiosample: np.array):
         return amp * signal
 
 
-class Drum:
+class Synth:
+    """
+    An abstract class for a modular synth, ensuring that all modules
+    have the same sample and control rate.
+    """
+
+    def __init__(self, modules: List[SynthModule]):
+        # Check that we are not mixing different control rates or sample rates
+        for m in modules[:1]:
+            assert m.sample_rate == modules[0].sample_rate
+            assert m.control_rate == modules[0].control_rate
+
+
+class Drum(Synth):
     """
     A package of modules that makes one drum hit.
     """
 
     def __init__(
         self,
+        sustain_duration: float,
         pitch_adsr: ADSR = ADSR(),
         amp_adsr: ADSR = ADSR(),
         vco: VCO = VCO(),
         vca: VCA = VCA(),
-        sustain_duration: float = 0,
     ):
+        super().__init__(modules=[pitch_adsr, amp_adsr, vco, vca])
+        assert sustain_duration >= 0
 
+        # TODO: Do we want the same sustain_duration for both ADSRs?
```

Comment:
That's the convention for triggering a note event, yes.

Reply:
Updated comment